src/llfs/ioring_log_driver.ipp

```diff
@@ -395,45 +426,69 @@ inline void BasicIoRingLogDriver<FlushOpImpl>::poll_commit_state()
 
 //==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
 //
-template <template <typename> class FlushOpImpl>
-inline void BasicIoRingLogDriver<FlushOpImpl>::wait_for_commit(slot_offset_type least_upper_bound)
+template <template <typename> class FlushOpImpl, typename StorageT>
+inline void BasicIoRingLogDriver<FlushOpImpl, StorageT>::wait_for_commit(
+    slot_offset_type least_upper_bound)
 {
   this->waiting_for_commit_.push(least_upper_bound);
   this->poll_commit_state();
 }
 
 //==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
 //
-template <template <typename> class FlushOpImpl>
-inline void BasicIoRingLogDriver<FlushOpImpl>::handle_commit_pos_update(
+template <template <typename> class FlushOpImpl, typename StorageT>
+inline void BasicIoRingLogDriver<FlushOpImpl, StorageT>::handle_commit_pos_update(
     const StatusOr<slot_offset_type>& updated_commit_pos)
 {
   // VERY IMPORTANT: we must execute the actual handler logic on the ioring thread to avoid data
   // races!
   //
-  this->ioring_.post(make_custom_alloc_handler(  //
-      this->commit_handler_memory_,              //
+  this->storage_.post_to_event_loop(make_custom_alloc_handler(  //
+      this->commit_handler_memory_,                             //
       [this, updated_commit_pos](const StatusOr<i32>& post_result) {
+        LLFS_VLOG(2) << "(driver=" << this->name_ << ") handle_commit_pos_update("
+                     << updated_commit_pos << ")" << BATT_INSPECT(post_result);
+
         this->commit_pos_listener_active_ = false;
 
         if (!updated_commit_pos.ok() || !post_result.ok()) {
-          this->ioring_.on_work_finished();
-          this->ioring_.stop();
+          this->storage_work_finished();
           return;
         }
 
-        LLFS_VLOG(2) << "(driver=" << this->name_
-                     << ") commit_pos listener invoked: " << updated_commit_pos;
-
         this->poll_commit_state();
       }));
 }
 
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+template <template <typename> class FlushOpImpl, typename StorageT>
+inline void BasicIoRingLogDriver<FlushOpImpl, StorageT>::storage_work_started()
+{
+  const bool was_working = this->storage_working_.exchange(true);
+  if (!was_working) {
+    this->storage_.on_work_started();
+  } else {
+    LLFS_LOG_WARNING() << "storage started again?" << std::endl << boost::stacktrace::stacktrace{};
+  }
+}
+
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+template <template <typename> class FlushOpImpl, typename StorageT>
+inline void BasicIoRingLogDriver<FlushOpImpl, StorageT>::storage_work_finished()
+{
+  const bool was_working = this->storage_working_.exchange(false);
+  if (was_working) {
+    this->storage_.on_work_finished();
+  }
```

Comment:
This is a bit different from `storage_work_started()` because whereas in that case there is really only one legit place it should be called, here there are a couple different sites that could be the first to call, depending on why we are shutting down, and we only want to call `on_work_finished()` the first time.